src/math/lp/lp_bound_propagator.h

```diff
@@ -83,79 +108,251 @@ class lp_bound_propagator {
         ~reset_cheap_eq() { p.reset_cheap_eq_eh(); }
     };
 
-   public:
+public:
     lp_bound_propagator(T& imp) : m_imp(imp) {}
 
-    const vector<implied_bound>& ibounds() const { return m_ibounds; }
+    const std_vector<implied_bound>& ibounds() const { return m_ibounds; }
 
     void init() {
-        m_improved_upper_bounds.clear();
-        m_improved_lower_bounds.clear();
-        m_ibounds.reset();
+        m_improved_upper_bounds.reset();
+        m_improved_lower_bounds.reset();
+        m_ibounds.clear();
         m_column_types = &lp().get_column_types();
     }
+   
+    bool is_linear(const svector<lpvar>& m, lpvar& zero_var, lpvar& non_fixed) {
+        zero_var = non_fixed = null_lpvar;
+        unsigned n_of_non_fixed = 0;
+        for (lpvar v : m) {
+            if (!this->column_is_fixed(v)) {
+                n_of_non_fixed++;
+                non_fixed = v;
+                continue;                
+            } 
+            const auto & b = get_lower_bound(v).x;
+            if (b.is_zero()) {
+                zero_var = v;
+                return true;
+            } 
+            
+        }
+        return n_of_non_fixed <= 1;
+    }
 
-    const lar_solver& lp() const { return m_imp.lp(); }
-    lar_solver& lp() { return m_imp.lp(); }
+    void add_bounds_for_zero_var(lpvar monic_var, lpvar zero_var) {
+        auto lambda = [zero_var](int* s) {
```

Comment:
type casting should not be necessary. It looks like they all just access lp() and not the lp_bound_propagator.
A virtual base class could also be used instead of a lambda.

Reply:
I could not use the lar_solver type in lp_bound_propagator because of the include dependencies. Notice that only lar_solver& is used there. "A virtual base class could also be used instead of a lambda": do you mean deriving from implied_bound on override explain_bound?